src/eos-util.c

```diff
@@ -0,0 +1,423 @@
+  const gchar *msg = (error && error->message) ? error->message : "Unspecified";
+
+  eos_updater_set_error_code (updater, code);
+  eos_updater_set_error_message (updater, msg);
+  eos_updater_set_state_changed (updater, EOS_UPDATER_STATE_ERROR);
+}
+
+OstreeDeployment *
+eos_updater_get_booted_deployment_from_loaded_sysroot (OstreeSysroot *sysroot,
+                                                       GError **error)
+{
+  OstreeDeployment *booted_deployment = NULL;
+
+  booted_deployment = ostree_sysroot_get_booted_deployment (sysroot);
+  if (booted_deployment == NULL)
+    {
+      g_set_error_literal (error, G_IO_ERROR, G_IO_ERROR_NOT_FOUND,
+                           "Not an ostree system");
+      return NULL;
+    }
+
+  return g_object_ref (booted_deployment);
+}
+
+OstreeDeployment *
+eos_updater_get_booted_deployment (GError **error)
+{
+  g_autoptr(OstreeSysroot) sysroot = ostree_sysroot_new_default ();
+
+  if (!ostree_sysroot_load (sysroot, NULL, error))
+    return NULL;
+
+  return eos_updater_get_booted_deployment_from_loaded_sysroot (sysroot, error);
+}
+
+gchar *
+eos_updater_get_booted_checksum (GError **error)
+{
+  g_autoptr(OstreeDeployment) booted_deployment = NULL;
+
+  booted_deployment = eos_updater_get_booted_deployment (error);
+  if (booted_deployment == NULL)
+    return NULL;
+
+  return g_strdup (ostree_deployment_get_csum (booted_deployment));
+}
+
+gchar *
+eos_updater_get_baseurl (OstreeDeployment *booted_deployment,
+                         OstreeRepo *repo,
+                         GError **error)
+{
+  const gchar *osname;
+  g_autofree gchar *url = NULL;
+
+  osname = ostree_deployment_get_osname (booted_deployment);
+  if (!ostree_repo_remote_get_url (repo, osname, &url, error))
+    return NULL;
+
+  return g_steal_pointer (&url);
+}
+
+gboolean
+eos_updater_get_ostree_path (OstreeRepo *repo,
+                             gchar **ostree_path,
+                             GError **error)
+{
+  g_autoptr(OstreeDeployment) deployment = NULL;
+  g_autofree gchar *ostree_url = NULL;
+  g_autoptr(SoupURI) uri = NULL;
+  g_autofree gchar *path = NULL;
+  gsize to_move = 0;
+
+  deployment = eos_updater_get_booted_deployment (error);
+  if (deployment == NULL)
+    return FALSE;
+
+  ostree_url = eos_updater_get_baseurl (deployment,
+                                        repo,
+                                        error);
+  if (ostree_url == NULL)
+    return FALSE;
+
+  uri = soup_uri_new (ostree_url);
+  if (uri == NULL)
+    {
+      g_set_error (error,
+                   G_IO_ERROR,
+                   G_IO_ERROR_INVALID_DATA,
+                   "ostree %s remote's URL is invalid (%s)",
+                   ostree_deployment_get_osname (deployment),
+                   ostree_url);
+      return FALSE;
+    }
+
+  path = g_strdup (soup_uri_get_path (uri));
+  while (path[to_move] == '/')
+    ++to_move;
+  if (to_move > 0)
+    memmove (path, path + to_move, strlen (path));
```

Comment:
As far as I can tell, this puts the `NUL` byte in the right place (since it will be moved along with the rest of the string), but there will be undefined garbage afterwards - maybe it might be limiting the length to `strlen (path) - to_move`